about/yocto-wpe.md

```diff
@@ -0,0 +1,112 @@
+---
+layout: post
+title: "Try WPE with Yocto images for Raspberry Pi"
+tags: [about]
+data: { dateless: "true" }
+permalink: /about/yocto-wpe.html
+---
+
+One way to easily test WPE is to use Yocto images for Raspberry Pi
+devices, write them in a micro-SD card and use them to boot the
+device.
+
+
+## Getting and flashing the image on a micro-SD card
+
+Use the following values:
+* **MACHINE**: one of **{raspberrypi3-mesa|raspberrypi4-64-mesa}**
+  * raspberrypi3-mesa is 32-bit, to be able to test WPE in a more resource-starved environment
+  * raspberrypi4-64-mesa is 64-bit
+* **SDIMG**: based on the machine, '`core-image-weston-browsers-${MACHINE}.wic`', e.g. '`core-image-weston-browsers-raspberrypi3-mesa.wic`'
+* **SDCARD_DEV**: whatever the device in your system, e.g. '`/dev/mmcblk0_REMOVE_THIS`'
+  * **note that PREVIOUS DATA WILL BE ERASED**, be specially careful to not overwrite the wrong disk/card
+
+The version of WPE in these images is **2.38.x**.
+
+
+### Getting the image
+
+Generically, using the substitutions mentioned previously:
+
+```
+wget https://wk-contrib.igalia.com/yocto/meta-webkit/browsers/stable/images/"${MACHINE}"/"${SDIMG}".{bz2,bmap}
+```
+
+### Writing using '`bmaptool`' (faster, recommended) or classic '`dd`'
+
+Homepage for bmaptool: [https://github.com/intel/bmap-tools](https://github.com/intel/bmap-tools)
+
+You can typically get this in GNU/Linux-based distributions installing '`bmap-tools`' package if available.
+
+
+### Combined instructions for Raspberry Pi 3B/3B+ (32-bit) and Raspberry Pi 4 (64-bit)
+
+Choose one of:
+```
+MACHINE=raspberrypi3-mesa
+```
+or:
+```
+MACHINE=raspberrypi4-64-mesa
+```
+
+Then to download the image:
+```
+SDIMG="core-image-weston-browsers-${MACHINE}.wic"
+SDCARD_DEV=/dev/mmcblk0_REMOVE_THIS
+wget https://wk-contrib.igalia.com/yocto/meta-webkit/browsers/stable/images/${MACHINE}/${SDIMG}.{bz2,bmap}
```

Comment:
I think this won't work unless you either add an `export` to the beginning of the first two lines, or put the three lines in one single command line, separated by spaces.

If, in the other hand, what you mean is that the user needs to do the substitutions manually, then I think it would be better to simply provide the two lines for 32bit and 64bit images, as I mentioned above.

Reply:
> Also, I think it would be worth running this by @bkardell or @meyerweb if possible too before merging, for an extra sanity check on possible language-related bits that could perhaps be improved.

@bkardell is added as reviewer (by default I think) but I'll make sure to ping him.

Reply:
This works fine without `export`, `exports` are only needed for children processes or something, but not when executing `wget` immediately:
```
$ MACHINE=raspberrypi3-mesa
$ SDIMG="core-image-weston-browsers-${MACHINE}.wic"
$ wget -nv https://wk-contrib.igalia.com/yocto/meta-webkit/browsers/stable/images/${MACHINE}/${SDIMG}.{bz2,bmap}
2023-01-10 11:06:34 URL:https://wk-contrib.igalia.com/yocto/meta-webkit/browsers/stable/images/raspberrypi3-mesa/core-image-weston-browsers-raspberrypi3-mesa.wic.bz2 [374408403/374408403] -> "core-image-weston-browsers-raspberrypi3-mesa.wic.bz2" [1]
2023-01-10 11:06:34 URL:https://wk-contrib.igalia.com/yocto/meta-webkit/browsers/stable/images/raspberrypi3-mesa/core-image-weston-browsers-raspberrypi3-mesa.wic.bmap [3506/3506] -> "core-image-weston-browsers-raspberrypi3-mesa.wic.bmap" [1]
FINISHED --2023-01-10 11:06:34--
Total wall clock time: 27s
Downloaded: 2 files, 357M in 27s (13.2 MB/s)
```

Anyway, as explained above: this was maybe more useful when having more machines like with the Wandboard that I had initially, and maybe in the future becomes relevant again, but at the moment is probably better to remove this part.

Reply:
> This works fine without `export`, `exports` are only needed for children processes or something, but not when executing `wget` immediately:

Oh! Sorry for the misleading comment... I was completely convinced an export would be needed, TIL...

Reply:
> So I think that in this part, specially with future considerations in mind, we should preserve the variables, WDYT?

I guess it works, especially now I realized `export` is not needed (:facepalm: ) but I'm still not very excited about having this, I'd rather have a link to the actual files as we only need two at the moment.

And in the future, if we have more images we can perhaps think of something slightly better and more scalable, like a bunch of combo boxes + some JavaScript that allows you to generate the download link and/or the snippet you need to use to get the images on the fly (e.g. something like what you can see in nvidia's website at https://www.nvidia.com/download/index.aspx, but simpler).